kernel/src/expressions/scalars.rs

```diff
@@ -83,6 +95,62 @@ impl ArrayData {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct MapData {
+    data_type: MapType,
+    pairs: Vec<(Scalar, Scalar)>,
+}
+
+impl MapData {
+    pub fn try_new(
+        data_type: MapType,
+        values: impl IntoIterator<Item = (impl Into<Scalar>, impl Into<Scalar>)>,
+    ) -> DeltaResult<Self> {
+        if let (key_type @ DataType::Primitive(_), val_type @ DataType::Primitive(_)) =
+            (data_type.key_type(), data_type.value_type())
+        {
+            let pairs = values
```

Comment:
Simpler (and less indentation)?
```suggestion
        let (key_type @ DataType::Primitive(_), val_type @ DataType::Primitive(_)) =
            (data_type.key_type(), data_type.value_type()) else
        {
            return Err(Error::unsupported(
                "Unable to construct Scalar MapData: map keys/values must be primitive types",
            ))
        };
        let pairs = values
```
Also, is there any particular reason we limit the map to primitive keys and values? The `Scalar::to_array` shouldn't care in the slightest.

Reply:
yea good catch - I think I remembered that in a couple places we treat the inside of maps/arrays as leaves (like `LiteralExpressionTransform` IIRC) but i do agree we should be able to just have Scalar represent complex types. i'll include a test as well